### DISCUSSION TITLE
Improve ability to check when shopping lists change

### DISCUSSION THREAD

**Discussion by xxxserxxx** [mealie-recipes/mealie#3449] [Feature Request]
### First Check

- [X] I used the GitHub search to find a similar requests and didn't find it.
- [X] Checked the [tasks tagged](https://github.com/mealie-recipes/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

### Please provide a concise description of the problem that would be addressed by this feature.

In the new API, shopping lists and list summaries have creation and modification times; these attributes only refer to changes to the attributes on the data structures themselves; that is, the modification time is only changed when e.g. the `name` is changed. The modification times are _not_ updated when list contents are changed; this means that it is impossible to tell if a shopping list has substantively changed by looking at a `listSummary`. Having a change timestamp on list summaries (and lists) would allow clients to more easily detect when shopping list _contents_ change -- and these are the sorts of changes in which humans are most interested.

### Please provide a concise description of the feature that would resolve your issue.

Ideally, `updateAt` would be changed when any change to anything in the list's `listItems` happens -- essentially, bubbling-up any change timestamps to the parent `list` and `listSummary` objects. However, the current timestamps are almost certainly a reflection of the serialization of the DBO, and so changing the behavior of `updateAt` attributes is unlikely. Therefore, I'd like a new attribute added to both `listSummary` and `list` that is updated when any change is made to a list's `listItems` contents or any item in the list.

### Please consider and list out some caveats or tradeoffs made in your design decision

In a shopping list, what is most significant to clients is the contents of the list. Changes to the contents -- items being added or removed, or renamed or checked-off -- are going to be the most common and most interesting changes, and it's currently hard to detect these changes: it's impossible from `listSummary`, and requires recursive timestamp checks. Since we're interested in when attributes at the leaves change -- a bottom-up change -- and because list trees are not deep, it should not be intrusive on the server side to update parent objects when their children change.

I am currently exclusively concerned about shopping lists; impacts on recipes or other structures have been excluded.

This change would significantly reduce client calls and data transfer for detecting changes interesting to humans.

I'm interested in this feature as the author of an OSS shopping list client that Mealie users can use on mobile devices and while offline.

### Additional Information

- [ ] If this is accepted I'm willing to submit a PR to provide this feature
- [ ] If this is accepted I'm willing to help maintain this feature
- [ ] I'm willing to sponsor/pay a developer to do this work

**Comment by michael-genson** [comment]
> Having a change timestamp on list summaries (and lists) would allow clients to more easily detect when shopping list contents change

> This change would significantly reduce client calls

Have you looked at our notifiers? You don't need to poll Mealie for changes, Mealie can send out notifications any time a shopping list item is added, updated, deleted, etc. This is how I built my shopping list integration with Alexa and Todoist: https://github.com/michael-genson/Unified-Shopping-List. If you're able to use a subscriber/consumer model I strongly recommend that over a polling model.

However, I still agree, the `updateAt` of the shopping list should reflect list item changes, everything you mention makes sense. I also understand not all use-cases are solved by the subscriber/consumer model.

**Comment by xxxserxxx** [comment]
The mobile app has to be able to work offline. It allows edits while offline and needs to be able to merge changes when the client comes online. The ability to work offline and to merge changes is a key reason why I wrote the app -- that, and to have a native client. It's a different operating model from an always-on application like an Alexa integration.

I do plan on using the pub/sub API, but on an optional side-loaded push notification server.  Push notifications are going to be more reliable on mobile than a bespoke pub/sub API -- again, simply because the client is mobile. This won't eliminate the need for the client to compare and merge list contents.

**Comment by michael-genson** [comment]
FYI this was implemented in https://github.com/mealie-recipes/mealie/pull/3453

**Comment by michael-genson** [comment]
Additionally, we're adding offline shopping list support here: https://github.com/mealie-recipes/mealie/pull/3760
